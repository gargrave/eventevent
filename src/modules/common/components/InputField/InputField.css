:root {
  --errorColor: #ff4136;
}

.errorMsg {
  color: var(--errorColor);
}

.invalid {
  border-color: var(--errorColor) !important; /* using !important to override Shoelace's styles */
}
