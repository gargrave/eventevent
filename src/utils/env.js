const ENV = {
  DEV: 'development',
  PROD: 'production',
};
const DEV_ENV_URL = 'localhost:3000';

let currentEnv;
(function setCurrentEnv() {
  if (!currentEnv) {
    const loc = window.location.href;
    if (loc.match(DEV_ENV_URL)) {
      currentEnv = ENV.DEV;
    } else {
      currentEnv = ENV.PROD;
    }
  }
})();

export const isDev = () => currentEnv === ENV.DEV;
export const isProd = () => currentEnv === ENV.PROD;
