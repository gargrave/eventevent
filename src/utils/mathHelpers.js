export const clamp = (val, min, max) => {
  if (val < min) {
    return min;
  }
  if (val > max) {
    return max;
  }
  return val;
};
