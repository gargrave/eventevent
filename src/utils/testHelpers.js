import React from 'react';
import { mount, shallow } from 'enzyme';

export class ComponentBuilder {
  constructor(component, defaultProps) {
    this.component = component;
    this.defaultProps = { ...defaultProps };
  }

  shallowGetComponent(extraProps = {}) {
    const props = {
      ...this.defaultProps,
      ...extraProps,
    };
    return shallow(<this.component {...props} />);
  }

  mountComponent(extraProps = {}) {
    const props = {
      ...this.defaultProps,
      ...extraProps,
    };
    return mount(<this.component {...props} />);
  }
}
