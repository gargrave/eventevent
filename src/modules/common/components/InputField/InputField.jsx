// @flow
import React from 'react';
import { bool, func, number, oneOf, string } from 'prop-types';

import { clamp } from '../../../../utils/mathHelpers';

import styles from './InputField.css';

const acceptableTypes = [
  'email',
  'password',
  'search',
  'text',
];

type Props = {
  boundValue: string,
  disabled?: boolean,
  error?: string,
  label?: string,
  maxLength?: number,
  name: string,
  onInputChange: Function,
  placeholder?: string,
  type?: string,
};

const InputField = ({
  boundValue,
  disabled,
  error,
  label,
  maxLength = 255,
  name,
  onInputChange,
  placeholder,
  type,
}: Props) => {
  return (
    <div className={'input-field'}>
      {label && <label htmlFor={name}>{label}:</label>}

      <input
        className={!!error ? styles.invalid : ''}
        disabled={disabled || false}
        id={name}
        name={name}
        maxLength={clamp(maxLength, 1, 255)}
        onChange={onInputChange}
        placeholder={placeholder || ''}
        type={type || 'text'}
        value={boundValue}
      />

      {error &&
        <p className={styles.errorMsg}>
          {error}
        </p>
      }
    </div>
  );
};

InputField.propTypes = {
  boundValue: string.isRequired,
  disabled: bool,
  error: string,
  maxLength: number,
  label: string,
  name: string.isRequired,
  onInputChange: func.isRequired,
  placeholder: string,
  type: oneOf(acceptableTypes),
};

export default InputField;
