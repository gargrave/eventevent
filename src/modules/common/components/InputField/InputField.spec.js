import { ComponentBuilder } from '../../../../utils/testHelpers';
import InputField from './InputField';

const defaultProps = () => ({
  boundValue: '',
  disabled: false,
  error: '',
  label: 'Whatever',
  name: 'text',
  onInputChange: jest.fn(),
  placeholder: '',
  type: 'text',
});

const builder = new ComponentBuilder(InputField, defaultProps());

describe('InputField', () => {
  let component;

  it('matches the snapshot', () => {
    component = builder.shallowGetComponent();
    expect(component).toMatchSnapshot();
  });

  describe('label display', () => {
    it('renders the correct label text', () => {
      const props = { label: 'test_label' };
      component = builder.shallowGetComponent(props);
      const label = component.find('label');
      expect(label).toHaveLength(1);
      expect(label.text()).toMatch(new RegExp(props.label));
    });

    it('does not render a label if prop is empty', () => {
      component = builder.shallowGetComponent({ label: '' });
      expect(component.find('label')).toHaveLength(0);
    });
  });

  describe('error display', () => {
    const errClass = '.error';

    it('renders the error correctly when one is present', () => {
      const errMsg = 'error_message';
      component = builder.shallowGetComponent({ error: errMsg });
      const err = component.find('p');
      expect(err).toHaveLength(1);
      expect(err.text()).toMatch(new RegExp(errMsg));
    });

    it('does not render an error if prop is empty', () => {
      component = builder.shallowGetComponent();
      expect(component.find(errClass)).toHaveLength(0);
    });
  });

  describe('"text" type input field', () => {
    it('renders a text input correctly', () => {
      component = builder.shallowGetComponent();
      expect(component.find('input[type="email"]')).toHaveLength(0);
      expect(component.find('input[type="text"]')).toHaveLength(1);
      expect(component.find('input[type="password"]')).toHaveLength(0);
    });
  });

  describe('"password" type input field', () => {
    it('renders a password input correctly', () => {
      component = builder.shallowGetComponent({ type: 'password' });
      expect(component.find('input[type="email"]')).toHaveLength(0);
      expect(component.find('input[type="text"]')).toHaveLength(0);
      expect(component.find('input[type="password"]')).toHaveLength(1);
    });
  });

  describe('"email" type input field', () => {
    it('renders an email input correctly', () => {
      component = builder.shallowGetComponent({ type: 'email' });
      expect(component.find('input[type="email"]')).toHaveLength(1);
      expect(component.find('input[type="password"]')).toHaveLength(0);
      expect(component.find('input[type="text"]')).toHaveLength(0);
    });
  });

  describe('maxLength', () => {
    it('has a default "maxlength" when none is supplied', () => {
      component = builder.shallowGetComponent();
      const input = component.find('input');
      expect(input.prop('maxLength')).toEqual(255);
    });

    it('correctly applies the "maxLength" attirbute', () => {
      component = builder.shallowGetComponent({ maxLength: 50 });
      const input = component.find('input');
      expect(input.prop('maxLength')).toEqual(50);
    });

    it('clamps min "maxLength" with zero value', () => {
      component = builder.shallowGetComponent({ maxLength: 0 });
      const input = component.find('input');
      expect(input.prop('maxLength')).toEqual(1);
    });

    it('clamps min "maxLength" with negative value', () => {
      component = builder.shallowGetComponent({ maxLength: -10 });
      const input = component.find('input');
      expect(input.prop('maxLength')).toEqual(1);
    });

    it('clamps max "maxLength"', () => {
      component = builder.shallowGetComponent({ maxLength: 256 });
      const input = component.find('input');
      expect(input.prop('maxLength')).toEqual(255);
    });
  });
});
