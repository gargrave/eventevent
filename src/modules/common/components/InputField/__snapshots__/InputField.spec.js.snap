// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InputField matches the snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <InputField
    boundValue=""
    disabled={false}
    error=""
    label="Whatever"
    name="text"
    onInputChange={[Function]}
    placeholder=""
    type="text"
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <label
          htmlFor="text"
>
          Whatever
          :
</label>,
        <input
          className=""
          disabled={false}
          id="text"
          maxLength={255}
          name="text"
          onChange={[Function]}
          placeholder=""
          type="text"
          value=""
/>,
        "",
      ],
      "className": "input-field",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            "Whatever",
            ":",
          ],
          "htmlFor": "text",
        },
        "ref": null,
        "rendered": Array [
          "Whatever",
          ":",
        ],
        "type": "label",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "className": "",
          "disabled": false,
          "id": "text",
          "maxLength": 255,
          "name": "text",
          "onChange": [Function],
          "placeholder": "",
          "type": "text",
          "value": "",
        },
        "ref": null,
        "rendered": null,
        "type": "input",
      },
      "",
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <label
            htmlFor="text"
>
            Whatever
            :
</label>,
          <input
            className=""
            disabled={false}
            id="text"
            maxLength={255}
            name="text"
            onChange={[Function]}
            placeholder=""
            type="text"
            value=""
/>,
          "",
        ],
        "className": "input-field",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              "Whatever",
              ":",
            ],
            "htmlFor": "text",
          },
          "ref": null,
          "rendered": Array [
            "Whatever",
            ":",
          ],
          "type": "label",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "className": "",
            "disabled": false,
            "id": "text",
            "maxLength": 255,
            "name": "text",
            "onChange": [Function],
            "placeholder": "",
            "type": "text",
            "value": "",
          },
          "ref": null,
          "rendered": null,
          "type": "input",
        },
        "",
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
